p2panda-store/src/sqlite/models.rs

```diff
@@ -0,0 +1,73 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use std::convert::From;
+use std::str::FromStr;
+
+use p2panda_core::{Extensions, Hash, Header, PublicKey, RawOperation, Signature};
+use sqlx::FromRow;
+
+use crate::sqlite::store::deserialize_extensions;
+
+/// A single "raw" operation row as it is inserted in the database.
+#[derive(FromRow, Debug, Clone, PartialEq, Eq)]
+pub struct RawOperationRow {
+    hash: String,
+    pub body: Option<Vec<u8>>,
+    header_bytes: Vec<u8>,
+}
+
+/// A single operation row as it is inserted in the database.
+#[derive(FromRow, Debug, Clone, PartialEq, Eq)]
+pub struct OperationRow {
+    hash: String,
+    log_id: String,
+    version: String,
+    public_key: String,
+    signature: Option<String>,
+    payload_size: String,
+    payload_hash: Option<String>,
+    timestamp: String,
+    seq_num: String,
+    backlink: Option<String>,
+    previous: String,
+    extensions: Option<Vec<u8>>,
+    pub body: Option<Vec<u8>>,
+    header_bytes: Vec<u8>,
+}
+
+impl<E> From<OperationRow> for Header<E>
+where
+    E: Extensions,
+{
+    fn from(row: OperationRow) -> Self {
+        let mut row_previous = row.previous;
+        let mut previous = Vec::new();
+        while !row_previous.is_empty() {
+            let (hex, rest) = row_previous.split_at(32);
+            // We assume database values are valid and therefore we're safe to unwrap.
+            previous.push(Hash::from_str(hex).unwrap());
+            row_previous = rest.to_string();
+        }
+
+        Header {
+            version: row.version.parse::<u64>().unwrap(),
+            public_key: PublicKey::from_str(&row.public_key).unwrap(),
+            signature: row.signature.map(|hex| Signature::from_str(&hex).unwrap()),
```

Comment:
Following my previous comment, maybe we could match on the signature here and error if it is none.

Reply:
I guess this will require adding our own error type...

Reply:
Its fine to panic here, as itll be a bug I assume

Reply:
Good call!